stan/math/rev/fun/inverse.hpp

```diff
@@ -75,16 +22,21 @@ class inverse_vari : public vari {
  * size zero).
  * @throw std::invalid_argument if the matrix is not square.
  */
-inline matrix_v inverse(const matrix_v &m) {
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline plain_type_t<T> inverse(const T& m) {
   check_square("inverse", "m", m);
   if (m.size() == 0) {
-    return {};
+    return m;
   }
 
-  matrix_v res(m.rows(), m.cols());
-  internal::inverse_vari *baseVari = new internal::inverse_vari(m);
-  res.vi() = Eigen::Map<matrix_vi>(baseVari->vari_ref_A_inv_, res.rows(),
-                                   res.cols());
+  arena_t<plain_type_t<T>> arena_m = m;
+  arena_t<Eigen::MatrixXd> res_val = value_of(arena_m).inverse();
+  arena_t<plain_type_t<T>> res = res_val;
+
+  reverse_pass_callback([res, res_val, arena_m]() mutable {
+    arena_m.adj() -= res_val.transpose() * res.adj_op() * res_val.transpose();
+  });
```

Comment:
[optional]

I wonder if this is one of the places where casting `m` to a `var<matrix>` is always better. We pull out the values and read through them multiple times, then also have to read through the adjoints of `arena_m`. If we made this always use `var<matrix>` for arena_m we'd have to pay the cost of writing the adjoints from the `var<matrix>` back to the `matrix<var>`, but we would get nice access in the `reverse_pass_callback()` here

Reply:
These operations are memory bound. So it only matters how much data we need to transfer between memory and cpu. Assuming the matrix is large enough that it does not fit into cache:

arena_m has value accessed once and adjoint once. Each of those need to load pointers, values and adjoints in cache. Incrementing the adjoints also need to store values and adjoints back into ram. This in total transfers 8*64bits per element over the bus. 

Transforming to varmat (`to_var_value()`) copies pointers to arena(2), copies values(loading pointers, values and adjoints and storing values, in total 4) and in reverse pass increments adjoint (oading pointers, values and adjoints and storing adjoints and values, in total 5). Using this varmat still means loading values (1) and incrementing adjoints(2). In total that would mean 2+4+5+1+2=14*64bits per element of transfers over the bus, making this slower.